enhancements/baremetal/continuous-hw-introspection.md

```diff
@@ -0,0 +1,147 @@
+---
+title: Continuous hardware introspecton of Baremetal IPI hosts
+authors:
+  - "@sadasu"
+reviewers:
+  - "@dhellmann"
+  - "@enxebre"
+  - "@hardys"
+  - "@stbenjam"
+  - "@juliakreger"
+approvers:
+  - "@dhellmann"
+  - "@enxebre"
+creation-date: 2020-02-18
+last-updated: 2020-02-17
+status: provisional|implementable|implemented|deferred|rejected|withdrawn|replaced
+---
+
+# Continuous hardware introspection of Baremetal IPI hosts
+
+This proposal talks about how performing continuous hardware introspection on all master
+and worker nodes in a Baremetal IPI deployment would get us important data about baremetal
+hosts which can be used to monitor changes in hardware capabilities of the server.
+
+This enhancement will not result in a new CRD or API to be written but will detail
+data that would become availabe in the Status field of the BareMetalHost CRD.
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [X] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Continuous hardware inspection on all the master and worker nodes that are a part
```

Comment:
This proposal seems to be redundant with device plugins: https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/

As for detecting some kind of failures, that is really the domain of the node problem detector.

I'd like to see a real end-to-end usecase of what this is actually used for.

Reply:
Thanks for a pointer to the device plugins. A quick scan revealed that using device plugins would require us to write drivers to gather information about hardware currently plugged into the server. But, we want to leverage an existing service that already reports data back to metal3 pod.

Reply:
This proposal separates out the "infrastructure" pieces that are needed for the collection of data and we don't want its progress to be slowed down by the discussion around how the collected data is going to be presented. I am working on another proposal that details how this information collected via continuous hardware introspection would be integrated into the metric stack. Please stay tuned.

Reply:
please see my comment here too: https://github.com/openshift/enhancements/pull/229#discussion_r388508679

Reply:
My concerns are that some of this tooling does not sound specific to the bare metal usecase, and we should try to utilize the tooling already available.  There is a higher level discussion about the UX of the product rather than just the needs of the metal3 use case that needs to be had, and I don't see any of the people that need to be involved in that discussion here.

Reply:
@michaelgugino the point here is to populate the BareMetalHost with information about the hardware so that e.g. if a disk is hot-swapped or a server is physically modified in some way (e.g. upgrading RAM) or the external network configuration is changed, the latest state is reflected in the Host. (In the case of master nodes this is needed for it to contain any state at all.) This allows the administrator responsible for the bare-metal infrastructure to treat this as a source of truth, and that *is* very specific to bare metal.
It's not about providing scheduling information to applications, and it includes a lot of information that wouldn't be suitable for device plugins (such as serial numbers of disks).

Reply:
@zaneb is that specific to bare metal?  Someone could attach a new drive to a VM on any number of platforms.